udata_hydra/analysis.py

```diff
@@ -75,49 +74,76 @@ async def process_resource(check_id: int) -> None:
     # if not, let's see if we can infer a modifification date from headers
     change_analysis = change_analysis or await detect_resource_change_from_headers(url) or {}
 
-    tmp_file = None
-    sha1 = None
-    error = None
-    # TODO: those can be infered from header, we could skip download totally
-    mime_type = None
-    filesize = None
-
-    try:
-        tmp_file = await download_resource(url, headers)
-    except IOError:
-        error = "File too large to download"
-    else:
-        # Get file size
-        filesize = os.path.getsize(tmp_file.name)
-        # Get checksum
-        sha1 = compute_checksum_from_file(tmp_file.name)
-        # Check if checksum has been modified if we don't have other hints
-        change_analysis = change_analysis or await detect_resource_change_from_checksum(resource_id, sha1) or {}
-        # FIXME: this never seems to output text/csv, maybe override it later
-        mime_type = magic.from_file(tmp_file.name, mime=True)
-    finally:
-        if tmp_file:
-            os.remove(tmp_file.name)
+    # if not, let's download the file to get some hints and other infos
+    dl_analysis = {}
+    if not change_analysis:
+        tmp_file = None
+        try:
+            tmp_file = await download_resource(url, headers)
+        except IOError:
+            dl_analysis["analysis:error"] = "File too large to download"
+        else:
+            # Get file size
+            dl_analysis["analysis:filesize"] = os.path.getsize(tmp_file.name)
+            # Get checksum
+            dl_analysis["analysis:checksum"] = compute_checksum_from_file(tmp_file.name)
+            # Check if checksum has been modified if we don't have other hints
+            change_analysis = (
+                await detect_resource_change_from_checksum(resource_id, dl_analysis["analysis:checksum"])
+                or {}
+            )
+            # TODO: this never seems to output text/csv, maybe override it later
+            dl_analysis["analysis:mime-type"] = magic.from_file(tmp_file.name, mime=True)
+        finally:
+            if tmp_file:
+                os.remove(tmp_file.name)
+            await update_check(check_id, {
+                "checksum": dl_analysis.get("analysis:checksum"),
+                "analysis_error": dl_analysis.get("analysis:error"),
+                "filesize": dl_analysis.get("analysis:filesize"),
+                "mime_type": dl_analysis.get("analysis:mime-type"),
+            })
+
+    has_changed_over_time = await detect_has_changed_over_time(change_analysis, resource_id, check_id)
+
+    change_analysis.update(dl_analysis)
+    if has_changed_over_time or (is_first_check and change_analysis):
```

Comment:
Makes me think the wording of `change_analysis` may be a bit weird in the right part. Indeed, we're in the case of a first check but still care about "change" analysis (that also encapsulates dl analysis).

We could rename it to `last_modified_analysis` for example?

Additionally, we could have a third analysis object for readability?
`complete_analysis={**change_analysis, **dl_analysis}`

Reply:
Additionally or alternatively? I went for alternatively ;-)